### ISSUE TITLE
[Bug]: Cannot update snapshots when testing ts_project

### ISSUE THREAD

**Issue by nathanosdev** [aspect-build/rules_jest#78]
### What happened?

When I run `jest_test` with a `ts_project` as a target instead of source files, updating snapshots fails.
I have the following logs:
```
$ bazel run //lib:test_update_snapshots
INFO: Analyzed target //lib:test_update_snapshots (377 packages loaded, 7350 targets configured).
INFO: Found 1 target...
INFO: From JsRunBinary lib/__snapshots__/out:
 PASS  lib/src/say-hello.spec.js
  sayHello
    ✓ should match snapshot (2 ms)

 › 1 snapshot written.
Snapshot Summary
 › 1 snapshot written from 1 test suite.

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   1 written, 1 total
Time:        1.193 s
Ran all test suites.
Target //lib:test_update_snapshots up-to-date:
  bazel-bin/lib/test_update_snapshots_update.sh
INFO: Elapsed time: 7.288s, Critical Path: 5.50s
INFO: 2005 processes: 1362 internal, 3 linux-sandbox, 638 local, 2 worker.
INFO: Build completed successfully, 2005 total actions
INFO: Build completed successfully, 2005 total actions
Copying directory /home/****/.cache/bazel/_bazel_****/332a89fded1c3dae890fec05ef02de2c/execroot/__main__/bazel-out/k8-fastbuild/bin/lib/test_update_snapshots_update.sh.runfiles/__main__/lib/__snapshots__/out to lib/__snapshots__ in /home/nathan/programming/playground/bazel_jest_issue
cp: cannot stat '/home/****/.cache/bazel/_bazel_****/332a89fded1c3dae890fec05ef02de2c/execroot/__main__/bazel-out/k8-fastbuild/bin/lib/test_update_snapshots_update.sh.runfiles/__main__/lib/__snapshots__/out/*': No such file or directory
```

### Version

Development (host) and target OS/architectures:
WSL on Windows 11
64 bit

Output of `bazel --version`:
bazel 5.3.2

### How to reproduce

```shell
Repro: https://github.com/nathanosdev/bazel_jest_snapshot_repro
```

### Any other information?

Thank you for your great work on these rules so far! I really love the direction that Bazel is bringing frontend technology.

### Fund our work

- [x] Sponsor our open source work by donating a [bug bounty](https://opencollective.com/aspect-build/)

**Comment by jbedard** [comment]
Does this work outside bazel? I think the `__snapshots__` directory needs to be a subdirectory of the directory the test files are in.

Thanks for the reproduction repo 👍

**Comment by nathanosdev** [comment]
> Does this work outside bazel? I think the `__snapshots__` directory needs to be a subdirectory of the directory the test files are in.

Yes, it works if I convert the TS files to JS and run `npx jest` in the `lib` folder. The snapshots are created in `lib/src/__snapshots__`/. It also works if I use JS and remove `rules_ts`. It's only breaking once I add `rules_ts` to the pipeline.

**Comment by jbedard** [comment]
If I upgrade everything in your repo (rules_js+ts+jest) and `mkdir lib/src/__snapshots__` it seems to work for me...

**Comment by jbedard** [comment]
Note that bazel requires the `mkdir` unlike outside bazel, which might be why it was working for you with `npx`?

**Comment by nathanosdev** [comment]
What worked for you? Were the snapshots written to disk when running `bazel test //lib:test_update_snapshots`?
For me, the logs say that it works, but nothing is actually written to disk.

I've updated the repro: https://github.com/nathanosdev/bazel_jest_snapshot_repro

```
$ mkdir -p ./lib/src/__snapshots__/
$ bazel run //lib:test_update_snapshots # looks like it works
$ bazel test //lib:test # still fails
```

As a sidenote, having to manually create `__snapshots__` is not an ideal DX. I hope this can be improved in a future version 👍

**Comment by gregmagolan** [comment]
If you set `snapshots = ["__snapshots__"]` instead of `snapshots = True` then the manual creation is not necessary. This could use some better documentation. It is mentioned, https://docs.aspect.build/rules/aspect_rules_jest/docs/rules#snapshots, but it could be made more clear.

The reason the creation is necessary with `True` is the hood that looks for all snapshot directories for that test target with `glob(["**/__snapshots__"])`: https://github.com/aspect-build/rules_jest/blob/3d76b1763b6326f31bb1ebb73a50b2f4b3087f05/jest/defs.bzl#L175

**Comment by nathanosdev** [comment]
I was already aware of that option, I think the documentation is clear on that, but it's not a nice transition for an existing project that already has 100s of components, each with their own `__snapshots__` folder to move to a single `__snapshots__` folder.

I understand there's some difficulties in this area anyway and Bazel is very picky sometimes, so I'm not complaining. Only hoping that it can be improved in the future :)

**Comment by gregmagolan** [comment]
Unfortunately, it is a hard requirement for Bazel to know all of the inputs & outputs in the graph during the analysis phase so it either needs to be told explicitly what the outputs are (which `__snapshots__` folder are created by the test) or it can attempt to discover it with `glob` but in that case the folders need to exist to be found. There is no way around that defined inputs/outputs requirements with Bazel as that is fundamental to the contract.

It could be improved in the future, however, with tooling that parses the specs and determines which once create snapshots and then updates the BUILD files accordingly.

**Comment by jbedard** [comment]
@nathanosdev your cli snippet works for me and I can't reproduce the original error🤔 

I simply did:
```
$ git clone git@github.com:nathanosdev/bazel_jest_snapshot_repro.git
$ cd bazel_jest_snapshot_repro
$ bazel test //...
$ mkdir -p ./lib/src/__snapshots__/
$ bazel run //lib:test_update_snapshots
$ bazel test //...
```

🤷

**Comment by jbedard** [comment]
System variables? Those shouldn't effect bazel unless they are declared inputs, so I'm not sure.

**Comment by nathanosdev** [comment]
I tried again on my other computer, wasn't working. I updated the latest Bazel, it worked. I rolled back to the previous version `5.3.2` and it was still working 🤷 So it looks like updating the rules fixes the issue, whatever other issue I had seems to be transient.

Thanks everyone 👏 I'll close the issue.
